Solving a one way ANOVA with Excel and testing for normality and homogeneity using Python (#5215)
section-engineering-education/engineering-education: content/articles/one-way-anova-excel-testing-normality-homogeneity-using-python/index.md

author: kuteyi-victor-toluwase
date: 2021-12-02T00:00:00-17:00
topics: []
excerpt_separator: <!--more-->
images:

  - url: /engineering-education/one-way-anova-excel-testing-normality-homogeneity-using-python/hero.jpg
    alt: One Way ANOVA with Excel using Python example image
---
Over the years, data has been an essential part of how the world works. These data range from GDP to blood samples to every aspect of the world. As data grew, statistics found ways to extract meaning from them. 
<!--more-->
One of these methods is the analysis of variance (ANOVA). ANOVA is a set of statistical models that analyses the differences among means. Microsoft came up with statistical add-ins within Excel to solve some of these problems. Although Microsoft Excel is still rather limited. Later came tools like R-studio and Python. 

With Python, the test for normality and homogeneity became easier. This tutorial will focus on creating an ANOVA table using Microsoft Excel, box plots to test the normality of a one way ANOVA, and Bartlett's rule for homogeneity with Python.

### Prerequisites
This tutorial requires readers to have proper knowledge on the following:

> - Here is a link to the [Analysis of variance (ANOVA)](https://www.investopedia.com/terms/a/anova.asp) documentation.

Reviewer: - Analysis of variance (ANOVA). Here's a [link]((https://www.investopedia.com/terms/a/anova.asp)) to the documentation.

- [Microsoft Excel](https://www.guru99.com/excel-tutorials.html).
- [Python](https://www.python.org/).

### Goal
At the end of the tutorial, the reader will be able to:
- Use Excel to create an ANOVA table.
- Install necessary Python dependencies to create a box plot.
- Test for normality using a box plot.
- Test for homogeneity using Bartlett's rule.

### Setting up the environment
To get started, we'll need to install Python and some of its dependencies. [Here](https://www.python.org/downloads/) is a link to download the latest Python executable based on your operating system. 

We need to install the following dependencies:
- pandas
- matplotlib
- seaborn

#### Installing pandas, matplotlib, seaborn
There are several methods to install Python dependencies, the most popular being the use of package managers like pip or conda. 

To install `pandas`, we'll need to:
- Open our terminal (either our ide's terminal or command prompt).

>Note: If you are using the command prompt, make sure to add the default folder to PATH.

- Type in the following command:
```bash
python -m pip install pandas
```

> 
> Wait a moment till the necessary dependencies are installed. 

Reviewer: ```suggestion
Wait a moment till the dependency is installed. 
```

Next, we'll follow the same steps for the other dependencies using the following commands respectively:
```bash
python -m pip install matplotlib
```

```bash
python -m pip install seaborn
```

After installing these dependencies, we'll proceed to create an ANOVA table with Excel.

### Getting started
#### Creating an ANOVA table with Excel
Microsoft Excels's Analysis Toolpak add-in is a game-changer in statistical Analysis. We'll show you how to create an ANOVA table containing the sum of squares, degree of freedom, mean squares, F value, P-value, F critical.

To perform a one way ANOVA, implement the following step.
- Import your data set in any preferred Excel format.

![data set](/engineering-education/Solving-a-one-way-anova-with-Excel-and-testing-for-normality-and-homogeneity-using-python/data-set.jpg)

> 

Reviewer: This is not the correct link for the image. You need to confirm the correct link from your files

- Check on Analysis Toolpak and select `Ok`.
- Return to the Data tab, and you'll notice the data analysis sub-tab.

#### Testing for normality using box plots
Now we will create a Python file and import pandas, matplotlib and seaborn into our code. Pandas, Matplotlib & Seaborn are Python libraries for data analysis. 

Pandas is used to import a wide variety of data formats, while Matplotlib and Seaborn are visualisation libraries used to generate powerful visualisations. Seaborn is a library that compiles statistical graphics in Python.

It uses the matplotlib library. Next, we'll read our dataset (it will be in CSV file format). Then we will reshape the data frame suitable for the statistical model package and replace the column names. Now we'll generate a boxplot to see the data distribution by treatments.

>Note: When reading from a directory, use double backslash or put `r` at the front of the directory address encased in a quote symbol.

![unicode error](/engineering-education/Solving-a-one-way-anova-with-Excel-and-testing-for-normality-and-homogeneity-using-python/unicode-error.jpg)

Below is an implementation of the above instructions:

> ```Python

Reviewer: Please change the 'P' in Python to be in small letters. Also, change it in the code snippet below.

Author: A content moderator @lalith1403 told me that I should make the first p in Python capital.

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
data_set = pd.read_csv(r"C:\Users\DELL\Desktop\Solving-a-one-way-anova-with-Excel-and-testing-for-normality-and-homogeneity-using-python\assignments.csv")
data_set_melt = pd.melt(data_set.reset_index(), id_vars=['index'], value_vars=['A', 'B', 'C', 'D', 'E'])
data_set_melt.columns = ['index', 'treatments', 'value']
ax = sns.boxplot(x='treatments', y='value', data=data_set_melt, color='#99c2a2')
ax = sns.swarmplot(x="treatments", y="value", data=data_set_melt, color='#7d0013')

> plt.show()

Reviewer: What's happening here in this code snippet? It's not enough to just write the code. You need to explain clearly to the readers.

Author: I made an explanation for the code snippet below in the text above.

```

On implementation, our box plot will be shown as below:

![boxplot](/engineering-education/Solving-a-one-way-anova-with-Excel-and-testing-for-normality-and-homogeneity-using-python/boxplot.jpg)

#### Testing for homogeneity using Bartlett's rule

> To do this, we'll import scipy.stats and call bartlett's method on our data set.

Reviewer: In a paragraph or two, you can begin by explaining to our readers what the Bartlett's rule entail before writing the code.

#### Example
```Python
A = 7,12,14,19,7
B = 7,17,18,25,10
C = 15,12,18,22,11
D = 11,18,19,19,15
E = 9,18,19,23,11
import scipy.stats as stats 
print(stats.bartlett(A, B, C, D, E))
```

When we run the above code on our terminal, we should get:

![bartlett](/engineering-education/Solving-a-one-way-anova-with-Excel-and-testing-for-normality-and-homogeneity-using-python/bartlett.jpg)

> 

Reviewer: Just like the other ones. Make sure your links are correct. And, no CAPS inside the link. Only small letters
